feat(Sym2): Add lemmas for equivSym

Jun2M opened the pull request:
There is a `equivSym` defined in "Data/Sym/Sym2.lean" with no lemma about what its properties are. This PR attempts to change that.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Jun2M committed:
equivSym lemmas

Jun2M committed:
Lint

j-loreaux committed:
feat: `le_of_sq_le_sq` (#18797)

loefflerd committed:
refactor(Analysis/Complex): merge Complex/Convex and Convex/Complex (#18819)

This merges the two files `Analysis/Complex/Convex.lean` and `Analysis/Convex/Complex.lean`, putting lemmas about convexity in the complex plane together; and moves a lemma from `UpperHalfPlane/Topology.lean` to this file, substantially shortening the import path to all the C-star-algebras files.

Whysoserioushah committed:
chore: use `Submodule.Quotient`-specific induction lemma (#18737)

joelriou committed:
feat(CategoryTheory/SmallObject/Iteration): the unique morphism (bot and succ cases) (#18127)




Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

j-loreaux committed:
feat: aliases of `IsSelfAdjoint.of_nonneg` for dot notation (#18798)

joelriou committed:
refactor(CategoryTheory/Monoidal): typeclasses Functor.LaxMonoidal, Functor.OplaxMonoidal and Functor.Monoidal (#17904)

Monoidal functors are refactored. Given a functor `F : C ⥤ D` between monoidal categories, we introduce typeclasses `F.LaxMonoidal` and `F.OplaxMonoidal` which carry the data of morphisms like `μ : F.obj X ⊗ F.obj Y ⟶ F.obj (X ⊗ Y)` (or `δ : F.obj (X ⊗ Y) ⟶ F.obj X ⊗ F.obj Y`). Then, the functor is monoidal if both `F.LaxMonoidal` and `F.OplaxMonoidal` and that both data provide inverse isomorphisms. With this design, there is more symmetry between lax and oplax functors (in the bundled setting, this was also experimented upon in #10845 ). (Roughly speaking, what was previously in `CategoryTheory.Monoidal.Functorial` was generalized.)

This is also a change from the previous designs where (lax) monoidal functors were bundled, which would be quite problematic or at least very unpractical in future applications: sooner or later, we shall have functors between derived categories, which are lax/oplax/monoidal, but are also triangulated, and some of these functors shall be adjoints, etc. The bundled aspect was not very much in use before this refactor: it was relevant only when constructing functors from a category of lax monoidal functors; in such case, the corresponding case of bundled lax monoidal functors, etc, have been kept in the code.



Co-authored-by: Joël Riou <joel.riou@universite-paris-saclay.fr>

zeramorphic committed:
feat: add Dockerfiles for lean and mathlib (#15936)

Add Dockerfiles for lean and mathlib



Co-authored-by: Kim Morrison <kim@tqft.net>

urkud committed:
feat(Order/Basic): add `gcongr` lemmas (#18476)

hrmacbeth committed:
feat: extend `linear_combination` to prove inequalities (#16841)

This PR extends the `linear_combination` tactic to prove inequalities as well as equalities.  From the test file:

```lean
example (a b : ℚ) (h1 : a ≤ 1) (h2 : b = 1) : (a + b) / 2 ≤ 1 := by
  linear_combination (h1 + h2) / 2

example {x y : ℤ} (hx : x + 3 ≤ 2) (hy : y + 2 * x ≥ 3) : y > 3 := by
  linear_combination hy + 2 * hx

example {x y z : ℚ} (h1 : 4 * x + y + 3 * z ≤ 25) (h2 : -x + 2 * y + z = 3)
    (h3 : 5 * x + 7 * z = 43) :
    x ≤ 4 := by
  linear_combination (14 * h1 - 7 * h2 - 5 * h3) / 38

example {t : ℚ} (ht : t ≥ 10) : t ^ 2 - 3 * t - 17 ≥ 5 := by
  linear_combination (t + 7) * ht
```

mathlib-bors[bot] committed:
chore: update Mathlib dependencies 2024-11-11 (#18832)

This PR updates the Mathlib dependencies.

j-loreaux committed:
chore: change `comm*TopologicalClosure` from `def` to `abbrev` (#18766)

jsm28 committed:
feat(Data/List): lemmas about `ofFn` (#18593)

Add several lemmas that proved convenient in working with `List.ofFn`.

There were two places where avoiding extra imports was awkward.  In the proof of `head_ofFn`, I wanted to rewrite with `head_eq_getElem_zero`, but `Mathlib.Data.List.Basic` isn't imported in `Mathlib.Data.List.OfFn` so I inlined an equivalent rewrite on `← getElem_zero (length_ofFn _ ▸ Nat.pos_of_ne_zero hn)` instead. `chain'_ofFn` depends on content from both `Mathlib.Data.List.OfFn` and `Mathlib.Data.List.Chain`, but neither of those imports the other. So to avoid adding extra imports or awkward local duplication of proofs, I created a new `Mathlib.Data.List.ChainOfFn` to have that one lemma.  Alternative suggestions for avoiding import creep for these lemmas are welcome.

Golfing is also welcome; the proof of `find?_ofFn_eq_some` is rather long.

Ruben-VandeVelde committed:
feat: Finite.algEquiv (#18825)

urkud commented on Mathlib/Data/Sym/Sym2.lean:
```diff
@@ -426,6 +426,12 @@ theorem IsDiag.mem_range_diag {z : Sym2 α} : IsDiag z → z ∈ Set.range (@dia
   ⟨IsDiag.mem_range_diag, fun ⟨i, hi⟩ => hi ▸ diag_isDiag i⟩
 
+@[simp]
+lemma map_IsDiag_iff (f : α ↪ β) (s : Sym2 α) :
```
I think that it's better to use `f : α → β` with `(hf : f.Injective)`. Also, please use implicit args.

urkud commented on Mathlib/Data/Sym/Sym2.lean:
```diff
@@ -566,6 +572,49 @@ def sym2EquivSym' : Equiv (Sym2 α) (Sym' α 2) where
+theorem equivSym_symm_eq_pair {a b : α} : (equivSym α).symm ⟨a ::ₘ {b}, rfl⟩ = s(a, b) := rfl
+
+@[simp]
+theorem mem_equivSym_iff_mem (s : Sym2 α) (a : α) : a ∈ equivSym α s ↔ a ∈ s := by
```
```suggestion
theorem mem_equivSym_iff_mem {s : Sym2 α} {a : α} : a ∈ equivSym α s ↔ a ∈ s := by
```

mathlib-bors[bot] committed:
chore: update Mathlib dependencies 2024-11-11 (#18834)

This PR updates the Mathlib dependencies.

urkud committed:
feat(Topology/Order): add more `CovBy.nhdsWithin...` lemmas (#18567)

- Add lemmas about `𝓝[<] a` and `𝓝[≤] a` in a `PredOrder`.
- Add lemmas about `𝓝[>] a` and `𝓝[≥] a` in a `SuccOrder`.
- Add `CovBy.nhdsWithin_Ici` and `CovBy.nhdsWithin_Iic`.
- Prove that an `OrderClosedTopology` on a linear order which is a `PredOrder` and a `SuccOrder` is discrete.
- Cleanup theorems about `OrderTopology` vs `DiscreteTopology`.

urkud committed:
feat(Topology/ContinuousMap): add `SecondCountableTopology` (#17706)

alreadydone committed:
chore: remove redundant imports from AlgebraicIndependent (#18791)

Command-Master committed:
doc(Topology/Order/LocalExtr): fix docstring for IsLocalExtr (#18843)

urkud committed:
feat(List/Basic): add `ext_getElem!` (#18802)

Also add `Sublist.of_cons_of_ne`.
Cherry-picked + golfed from #9082

erdOne committed:
feat(AlgebraicGeometry): quasi-compact + specializing => closed map (#18032)

Partially from the valuative criterion project

Co-authored-by: Christian Merten



Co-authored-by: Andrew Yang <36414270+erdOne@users.noreply.github.com>

jcommelin committed:
chore(workflows/discover-lean-pr-testing.yml): checkout mathlib4 instead of mathlib (#18840)

adomani committed:
feat: maintainer delegate everywhere (#18736)

The workflow file in this PR is meant to replace the three files for the maintainer merge from comment, review and review comment.

Besides having the same `maintainer merge` functionality, it also adds the `maintainer delegate` functionality.

I did not delete the old workflow files in the PR, since maybe, in order to properly test it, it may be a good idea to have the old and new workflows run in parallel for a final test.

In case you want to see the tests for the script, [this PR](https://github.com/bryangingechen/mathlib4/pull/14) has a live version of the workflow and you can test putting `maintainer merge/delegate` comments directly, from review or from review comment.

Note that the *test* workflow does *not* post to Zulip, nor does it check that who uses the maintainer merge feature is a reviewer.  Since the action does not post on Zulip, it posts the message instead in a PR comment.

acmepjz committed:
feat: add results on cardinality of localization (#18004)

- `OreLocalization.oreDiv_one_surjective_of_finite_(left|right)`: if `S` or `X` is finite, then the natural map `X → X[S⁻¹]` is surjective
- `OreLocalization.numeratorHom_surjective_of_finite`: if `S` is finite, then the natural map `R → R[S⁻¹]` is surjective
- `OreLocalization.cardinalMk_le_max`: `#X[S⁻¹] ≤ max(#S, #X)`
- `OreLocalization.cardinalMk_le`: `#R[S⁻¹] ≤ #R`
- `OreLocalization.cardinalMk_le_lift_cardinalMk_of_commute`: if `S` is commutative, then `#X[S⁻¹] ≤ #X` (I don't know how to remove the commutative assumption)
- `OreLocalization.cardinalMk`: if `S` has no zero divisors, then `#R[S⁻¹] = #R`

Also the `[lift_]cardinalMk_le` and `[lift_]cardinalMk` for `Localization`, `IsLocalization`, `FractionRing` and `IsFractionRing`.

Also change the direction of the conclusion of the existing `IsLocalization.card` (renamed to `IsLocalization.cardinalMk`) to match the docstring in file header. It is not directly used in any other files.

jcommelin committed:
chore(LongestPole): avoid importing file containing `main` (#18846)

TwoFX committed:
chore(CategoryTheory): dualize `IndYoneda` (#18585)



Co-authored-by: Markus Himmel <markus@lean-fro.org>

YaelDillies committed:
feat: Pontryagin duality for finite abelian groups (#15999)

Prove Pontryagin duality in case of finite abelian groups.

This states that any finite abelian group is canonically isomorphic to its double dual (the space of complex-valued characters of its space of complex-valued characters).

We first prove it for `ZMod n` and then extend to all finite abelian groups using the Structure Theorem.

From LeanAPAP

Co-authored-by Bhavik Mehta <bhavikmehta8@gmail.com>

b-mehta committed:
feat(InfiniteSum): infinite product of absolute value (#18747)



Co-authored-by: David Loeffler <d.loeffler.01@cantab.net>

Vierkantor committed:
chore: some `erw`s that can become `rw`s for free (#18847)

The linter of #17638 found five more `erw`s that could also be `rw`s. 🐙

Vierkantor committed:
chore(RingTheory/Noetherian): split `Noetherian.lean` (#18734)

This PR splits `RingTheory/Noetherian.lean` into the following files:
* `Noetherian/Defs.lean`: definition of `IsNoetherian` and `IsNoetherianRing`; proof of `isNoetherian_iff`
* `Noetherian/Filter.lean`: ascending chain condition, expressed in terms of filters
* `Noetherian/Orzech.lean`: Orzech's theorem and property
* `Noetherian/Nilpotent.lean`: nilpotent ideals in Noetherian rings
* `Noetherian/UniqueFactorization.lean`: Noetherian implies unique factorization

This change should also clean up the import graph for `LinearAlgebra` somewhat, especially in combination with PR #18731.

Note that there was a timeout in `Topology/Sheaves/Stalks.lean` caused by the instance search order changing. We can either do a local priority change, or simply avoid importing the slow instances. [See also Zulip thread](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Instance.20order.20sensitivity.20in.20.60Topology.2FSheaves.2FStalks.2Elean.60).



Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

jsm28 committed:
feat(Data/Nat/Nth): `count_nth_succ_of_infinite` (#18835)

Add a lemma `count_nth_succ_of_infinite`, which is to `count_nth_succ` as `count_nth_of_infinite` is to `count_nth`.

mathlib-bors[bot] committed:
chore: update Mathlib dependencies 2024-11-11 (#18848)

This PR updates the Mathlib dependencies.

erdOne committed:
refactor(AlgebraicGeometry): weaken `RingHom.PropertyIsLocal` (#18717)

Ruben-VandeVelde committed:
chore: redefine AlgEquiv.refl with RingEquiv.refl (#18856)

Using `1` requires an unnecessary import.

mathlib-bors[bot] committed:
chore: update Mathlib dependencies 2024-11-11 (#18859)

This PR updates the Mathlib dependencies.

hrmacbeth committed:
chore: change some `linarith`s to `omega`s (#18839)

Since `omega` is a lot faster than `linarith`, I assume we should prefer it to `linarith` in cases where both are applicable.

adomani committed:
fix(maintainer delegate): post in correct topic, add CI run link (#18855)

TwoFX committed:
chore: dualize Limits.Preserves.Opposites (#18814)

urkud committed:
feat(Data/Fin/Parity): add lemmas about parity of `Fin` numbers (#8960)

Co-authored-by: Iván Renison <85908989+IvanRenison@users.noreply.github.com>



Co-authored-by: Iván Renison <85908989+IvanRenison@users.noreply.github.com>

TwoFX committed:
feat(CategoryTheory): inclusion of terminal object is final (#18807)

Vierkantor committed:
fix(Algebra/Order/Field): fix instance diamond for `NNRat` (#18860)

The `Semifield ℚ≥0` instance used a default definition of `zpow` using `zpowRec`, while the `LinearOrderedSemifield` instance has a custom definition of `zpow` called `Nonneg.zpow` (using the fact that `zpow` preserves nonnegativity). These two are not defeq and therefore we found an instance diamond in #18830.

The fix is to use `Nonneg.zpow` (or something defeq to it) in the definition of `NNRat.instSemifield`. However, this is tricky because `Nonneg.zpow` is only defined after that instance.

I saw two ways to implement the fix:
 * postpone the `Semifield NNRat` instance to after the definition of `Nonneg.zpow`. It should definitely be possible to move all the `NNRat`-related definitions from `Algebra.Field.Rat` to `Algebra.Order.Field.Rat`, but it looks like the import graph gets a bit more tangled this way.
 * redo the proof of `Nonneg.zpow` inline in `NNRat.instSemifield`. It turns out `inv` and `div` are already redefined in the same way in the same section, so this seems like the neatest solution with the least surprises.

I also added an `example` that checks the diamond is indeed now fixed.

adomani committed:
chore: remove duplicated maintainer merge workflows (#18867)

These workflows are made obsolete by the new `maintainer_merge` script that implements all 3 in a single action.  This was the result of
* #18736, the initial PR;
* #18855, the fix for correct triaging of topics.

Vierkantor committed:
chore(Algebra/Field): split `Subfield.lean` into `Defs` and `Basic` (#18528)

This PR splits off a `Defs.lean` file from `Subfield.lean` containing on the definition and field structure of `Subfield`(`Class`).

This PR was motivated by the "hoard factor" computation that showed https://github.com/leanprover-community/mathlib4/pull/18520 increased the factor of `Mathlib.Algebra.Order.Field.Subfield` by a lot, meaning there are many transitive imports in that file whose declarations remain unused. The previous PR increased it from 0,814203 to 0,828997, this PR reduces it down to 0,784072

Vierkantor committed:
chore(Data/Finset): split `Basic.lean` into many smaller files (#18610)

The purpose of this PR is to turn the unwieldy file `Data/Finset/Basic.lean` into a collection of smaller files. The splits happen mostly based on the sections of the original `Basic.lean` files, so the import structure can probably be organized in a nicer way. There are also some results in `Basic.lean` that can find a nicer home if we organize the imports further. Still, I expect these changes help to clean up the file a lot.

The following new files are split off from `Basic.lean`:

* `Mathlib/Data/Finset/Attach.lean`: `Finset.attach` and elementary results
* `Mathlib/Data/Finset/Dedup.lean`: packages `Multiset.dedup` as `Multiset.toFinset`; some adjacent results for mapping between multisets, lists and finsets.
* `Mathlib/Data/Finset/Defs.lean`: definition of the `Finset` type and its membership predicates
* `Mathlib/Data/Finset/Disjoint.lean`: results on `Disjoint` applied to finsets; definition of `disjUnion`.
* `Mathlib/Data/Finset/Empty.lean`: definition of `∅` and the `Finset.Nonempty` predicate
* `Mathlib/Data/Finset/Erase.lean`: definition of `Finset.erase`
* `Mathlib/Data/Finset/Filter.lean`: definition of `Finset.filter` and the `{x ∈ s | p x}` set builder notation
* `Mathlib/Data/Finset/Insert.lean`: definition of `Finset.cons`, `Finset.erase` and the singleton finset `{a}`
* `Mathlib/Data/Finset/Lattice/Basic.lean`: definition of union and intersection of finsets, lattice structure.
* `Mathlib/Data/Finset/Lattice/Lemmas.lean`: results on union and intersection needed for the `GeneralizedBooleanAlgebra` instance in `SDiff.lean`
* `Mathlib/Data/Finset/Range.lean`: definition of `Finset.range`
* `Mathlib/Data/Finset/SDiff.lean`: definition of set difference on finsets, `GeneralizedBooleanAlgebra` instance
* `Mathlib/Data/Finset/SymmDiff.lean`: basic results on the symmetric difference operator (derived from set difference).

The following files have seen major changes:

* `Mathlib/Data/Finset/Basic.lean`: most of its contents have been removed. Still contains a few lesser-used definitions and an assortment of results that didn't have an obvious home at first sight.
* `Mathlib/Data/Finset/Lattice/Fold.lean`: renamed from `Mathlib/Data/Finset/Lattice.lean`

Vierkantor committed:
chore(Data/DFinsupp): split `Basic.lean` into many smaller files (#18656)

This PR splits `Data/DFinsupp/Basic.lean` into a collection of smaller files. The goal here is to keep the structure somewhat intact. Thankfully a somewhat clean split was possible while keeping imports reasonably small.

The following new files were made:
* `Data/DFinsupp/Defs.lean` has the definition of `DFinsupp` and all the essential definitions: `support`, `single`, `update`, the `AddCommMonoid` structure. (Or should this be renamed to `Basic.lean`?)
* `Data/DFinsupp/Ext.lean` has an extensionality principle for `AddMonoidHom`s composed with `DFinsupp`s. Might go into `Defs.lean` except it needs some heavy imports.
* `Data/DFinsupp/BigOperators.lean`: sums and products of `DFinsupp`s, and sums and products as `DFinsupp`s
* `Data/DFinsupp/FiniteInfinite.lean`: instances for `Fintype` and `Infinite`
* `Data/DFinsupp/Module.lean`: scalar multiplication and module structure; the `DFinsupp` maps as linear maps
* `Data/DFinsupp/Sigma.lean`: `sigmaCurry` and `sigmaUncurry`
* `Data/DFinsupp/Submonoid.lean`: lemmas involving submonoids, in particular that (additive) submonoids are closed under products (sums) of `DFinsupp`s

The following file no longer exists:
 * `Data/DFinsupp/Basic.lean` (split completely!)




Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

CBirkbeck committed:
feat(Analysis/Complex/AbsMax): add two lemmas (#18198)

add two results:

-  If `f` is differentiable on the open unit ball `{z : ℂ | ‖z‖ < 1}`, and `‖f‖` attains a maximum
in  this open ball, then `f` is constant 

-   If `f` is a function differentiable on the open unit ball, and there exists an `r < 1` such that
any value of `‖f‖` on the open ball is bounded above by some value on the closed ball of radius `r`,
then `f` is constant. 

These lemmas together with #18192 will be used to prove some dimension results for spaces of modular forms.

alreadydone committed:
feat: improve defeq in skeletonEquivalence (#18746)

+ Change the `inverse` in `skeletonEquivalence` to be defeq to `Quotient.mk`; it's previously obtained from `IsEquivalence.asEquivalence` by arbitrary choice.

+ Make `tensorIso` notation scoped (`tensorObj` and `tensorHom` are already scoped).

+ Add a `CoeSort` instance for `Sksleton`.

+ Add a few API lemmas.

alreadydone committed:
perf(Ideal/Quotient): directly use AddCommGroup instance in CommRing (#18824)

with a bigger effect when applied on top of #17930

chrisflav committed:
feat(AlgebraicGeometry): the big etale site (#18854)

For any morphism property `P`, we define the pretopology on the category of schemes defined by `P` as jointly surjective families of morphisms satisfying `P`. As an application, this is used to define the étale site.

The small site associated to `P` and with it the small étale site comes in a follow-up PR.

D-Thomine committed:
feat(Order.Filter.Basic): add frequently_congr (#18865)

YaelDillies committed:
feat: the translation operator (#16183)

Define translation of a function by an element of its domain

From LeanAPAP

adomani committed:
fix: silence the linter in files not imported by `Mathlib.lean` (#18033)

[Zulip](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/intrusive.20linters)

vihdzp committed:
chore(SetTheory/Ordinal/Basic): tweak preimage lemmas (#18061)

We move the misplaced `type_preimage` next to its generalization `type_lift_preimage`, and reinstantiate the latter's `simp` attribute.

fbarroero committed:
feat(Data/Int/WithZero): add API (#18668)

In this PR we add three basic facts about `WithZeroMultInt.toNNReal`.



Co-authored-by: mariainesdff <mariaines.dff@gmail.com>

Ruben-VandeVelde committed:
chore: split UFD results out of RingTheory.Polynomial.Basic (#18875)

riccardobrasca committed:
feat: module instance for quotient by `AddCommGroup.torsion` (#18060)

From flt-regular.

apnelson1 committed:
feat(Data/Matroid/Closure): Spanning sets in matroids (#18129)

We continue to build the API for matroid closure by defining a predicate `Matroid.Spanning` stating that a set `X` spans the ground set, or equivalently that `X` contains a base.

We also state a couple of dozen API lemmas for this notion.

b-mehta committed:
chore(SpecialFunctions/Log): transfer results from log to logb (#18621)

These are mostly transferred directly, except for `continuousAt_logb_iff` which needs different handling since it's not always true.

b-mehta committed:
chore(Algebra/InfiniteSum/Order): weaken typeclass assumptions (#18742)

Generalise the typeclass assumptions on three lemmas, with no other changes.

- Remove any assumptions relating the order and the algebra on `hasProd_le_of_prod_le`, also weaken OrderClosedTopology to ClosedIicTopology
- Remove any assumptions relating the order and the algebra on `le_hasProd_of_le_prod`, also weaken OrderClosedTopology to ClosedIciTopology
- Remove the T2 assumption on `tprod_le_of_prod_range_le`, also weaken OrderClosedTopology to ClosedIicTopology

The first two changes mean these lemmas can now be used for infinite products of reals.

sgouezel committed:
feat: add a lemma `Set.MapsTo.preimage_mem_nhdsWithin` (#18851)

Jun2M committed:
Update Mathlib/Data/Sym/Sym2.lean

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

alreadydone committed:
feat: tensoring a localization with itself does nothing (#18744)

`IsLocalization.tensorSelfAlgEquiv`: If A is a localization of a commutative ring R, then `A ⊗[R] A` is canonically isomorphic to `A` as `A`-algebras via multiplication. To construct the isomorphism we introduce a stronger version of `Algebra.TensorProduct.lmul'`.

Also add some easy theorem (`linearMap_compatibleSMul`) and defs (`distribBaseChange`, `LinearEquiv.baseChange`) and protect `IsLocalization.map_smul`.

erdOne committed:
feat(AlgebraicGeometry): the diagonal map is an immersion (#18726)

Also shows that injective on points => separated

partially from the valuation criterion project



Co-authored-by: Andrew Yang <36414270+erdOne@users.noreply.github.com>

vihdzp committed:
chore(Order/InitialSeg): `InitialSeg.ofIso` → `RelIso.toInitialSeg` (#18069)

This enables dot notation.

Jun2M committed:
Apply the suggested changes

YaelDillies committed:
feat: Incidence algebras (#8900)

Define incidence algebras

From LeanCamCombi

Co-authored-by: Alex J. Best <alex.j.best@gmail.com>

YaelDillies committed:
feat: product of sublattices (#17694)

From LeanCamCombi

YaelDillies committed:
chore(Topology): Namespace `Inducing`, `Embedding`... (#15993)

We don't add deprecated aliases because definitions only gained a namespace and the presence of both namespaced and unnamespaced declarations would confuse Lean. Furthermore, `_root_.IsInducing`, `_root_.IsEmbedding`,  etc... only existed for a short while.

Moves:
* `IsInducing` → `Topology.IsInducing`
* `IsEmbedding` → `Topology.IsEmbedding`
* `IsOpenEmbedding` → `Topology.IsOpenEmbedding`
* `IsClosedEmbedding` → `Topology.IsClosedEmbedding`
* `IsQuotientMap` → `Topology.IsQuotientMap`
* `IsClosedEmbedding.LindelofSpace` → `Topology.IsClosedEmbedding.LindelofSpace`
* `IsClosedEmbedding.compactSpace` → `Topology.IsClosedEmbedding.compactSpace`
* `IsClosedEmbedding.inl` → `Topology.IsClosedEmbedding.inl`
* `IsClosedEmbedding.inr` → `Topology.IsClosedEmbedding.inr`
* `IsClosedEmbedding.isCompact_preimage` → `Topology.IsClosedEmbedding.isCompact_preimage`
* `IsClosedEmbedding.isLindelof_preimage` → `Topology.IsClosedEmbedding.isLindelof_preimage`
* `IsClosedEmbedding.isProperMap` → `Topology.IsClosedEmbedding.isProperMap`
* `IsClosedEmbedding.locallyCompactSpace` → `Topology.IsClosedEmbedding.locallyCompactSpace`
* `IsClosedEmbedding.measurableEmbedding` → `Topology.IsClosedEmbedding.measurableEmbedding`
* `IsClosedEmbedding.nonLindelofSpace` → `Topology.IsClosedEmbedding.nonLindelofSpace`
* `IsClosedEmbedding.noncompactSpace` → `Topology.IsClosedEmbedding.noncompactSpace`
* `IsClosedEmbedding.normalSpace` → `Topology.IsClosedEmbedding.normalSpace`
* `IsClosedEmbedding.paracompactSpace` → `Topology.IsClosedEmbedding.paracompactSpace`
* `IsClosedEmbedding.preimage_closedPoints` → `Topology.IsClosedEmbedding.preimage_closedPoints`
* `IsClosedEmbedding.quasiSober` → `Topology.IsClosedEmbedding.quasiSober`
* `IsClosedEmbedding.restrictPreimage` → `Topology.IsClosedEmbedding.restrictPreimage`
* `IsClosedEmbedding.sigmaCompactSpace` → `Topology.IsClosedEmbedding.sigmaCompactSpace`
* `IsClosedEmbedding.sigmaMk` → `Topology.IsClosedEmbedding.sigmaMk`
* `IsClosedEmbedding.subtypeVal` → `Topology.IsClosedEmbedding.subtypeVal`
* `IsClosedEmbedding.t4Space` → `Topology.IsClosedEmbedding.t4Space`
* `IsClosedEmbedding.tendsto_coLindelof` → `Topology.IsClosedEmbedding.tendsto_coLindelof`
* `IsClosedEmbedding.tendsto_cocompact` → `Topology.IsClosedEmbedding.tendsto_cocompact`
* `IsClosedEmbedding.uliftDown` → `Topology.IsClosedEmbedding.uliftDown`
* `IsClosedEmbedding.weaklyLocallyCompactSpace` → `Topology.IsClosedEmbedding.weaklyLocallyCompactSpace`
* `IsEmbedding.codRestrict` → `Topology.IsEmbedding.codRestrict`
* `IsEmbedding.comapMetricSpace` → `Topology.IsEmbedding.comapMetricSpace`
* `IsEmbedding.comapUniformSpace` → `Topology.IsEmbedding.comapUniformSpace`
* `IsEmbedding.completelyNormalSpace` → `Topology.IsEmbedding.completelyNormalSpace`
* `IsEmbedding.continuousOn_iff` → `Topology.IsEmbedding.continuousOn_iff`
* `IsEmbedding.inclusion` → `Topology.IsEmbedding.inclusion`
* `IsEmbedding.inl` → `Topology.IsEmbedding.inl`
* `IsEmbedding.inr` → `Topology.IsEmbedding.inr`
* `IsEmbedding.isCompact_iff` → `Topology.IsEmbedding.isCompact_iff`
* `IsEmbedding.isLindelof_iff` → `Topology.IsEmbedding.isLindelof_iff`
* `IsEmbedding.isLocallyClosed_iff` → `Topology.IsEmbedding.isLocallyClosed_iff`
* `IsEmbedding.isSigmaCompact_iff` → `Topology.IsEmbedding.isSigmaCompact_iff`
* `IsEmbedding.isTotallyDisconnected` → `Topology.IsEmbedding.isTotallyDisconnected`
* `IsEmbedding.isTotallyDisconnected_image` → `Topology.IsEmbedding.isTotallyDisconnected_image`
* `IsEmbedding.isTotallyDisconnected_range` → `Topology.IsEmbedding.isTotallyDisconnected_range`
* `IsEmbedding.map_nhdsWithin_eq` → `Topology.IsEmbedding.map_nhdsWithin_eq`
* `IsEmbedding.measurableEmbedding` → `Topology.IsEmbedding.measurableEmbedding`
* `IsEmbedding.prodMap` → `Topology.IsEmbedding.prodMap`
* `IsEmbedding.restrictPreimage` → `Topology.IsEmbedding.restrictPreimage`
* `IsEmbedding.secondCountableTopology` → `Topology.IsEmbedding.secondCountableTopology`
* `IsEmbedding.separableSpace` → `Topology.IsEmbedding.separableSpace`
* `IsEmbedding.sigmaMk` → `Topology.IsEmbedding.sigmaMk`
* `IsEmbedding.subtypeVal` → `Topology.IsEmbedding.subtypeVal`
* `IsEmbedding.t0Space` → `Topology.IsEmbedding.t0Space`
* `IsEmbedding.t1Space` → `Topology.IsEmbedding.t1Space`
* `IsEmbedding.t25Space` → `Topology.IsEmbedding.t25Space`
* `IsEmbedding.t2Space` → `Topology.IsEmbedding.t2Space`
* `IsEmbedding.t3Space` → `Topology.IsEmbedding.t3Space`
* `IsEmbedding.t5Space` → `Topology.IsEmbedding.t5Space`
* `IsEmbedding.toPullbackDiag` → `Topology.IsEmbedding.toPullbackDiag`
* `IsEmbedding.to_isometry` → `Topology.IsEmbedding.to_isometry`
* `IsEmbedding.uliftDown` → `Topology.IsEmbedding.uliftDown`
* `IsInducing.alexandrovDiscrete` → `Topology.IsInducing.alexandrovDiscrete`
* `IsInducing.codRestrict` → `Topology.IsInducing.codRestrict`
* `IsInducing.comapPseudoMetricSpace` → `Topology.IsInducing.comapPseudoMetricSpace`
* `IsInducing.continuousConstSMul` → `Topology.IsInducing.continuousConstSMul`
* `IsInducing.continuousInv` → `Topology.IsInducing.continuousInv`
* `IsInducing.continuousMul` → `Topology.IsInducing.continuousMul`
* `IsInducing.continuousOn_iff` → `Topology.IsInducing.continuousOn_iff`
* `IsInducing.continuousSMul` → `Topology.IsInducing.continuousSMul`
* `IsInducing.continuousWithinAt_iff` → `Topology.IsInducing.continuousWithinAt_iff`
* `IsInducing.frechetUrysohnSpace` → `Topology.IsInducing.frechetUrysohnSpace`
* `IsInducing.generalizingMap` → `Topology.IsInducing.generalizingMap`
* `IsInducing.hasProd_iff` → `Topology.IsInducing.hasProd_iff`
* `IsInducing.injective` → `Topology.IsInducing.injective`
* `IsInducing.inseparable_iff` → `Topology.IsInducing.inseparable_iff`
* `IsInducing.isClosedMap` → `Topology.IsInducing.isClosedMap`
* `IsInducing.isCompact_iff` → `Topology.IsInducing.isCompact_iff`
* `IsInducing.isCompact_preimage` → `Topology.IsInducing.isCompact_preimage`
* `IsInducing.isCompact_preimage'` → `Topology.IsInducing.isCompact_preimage'`
* `IsInducing.isCompact_preimage_iff` → `Topology.IsInducing.isCompact_preimage_iff`
* `IsInducing.isEmbedding` → `Topology.IsInducing.isEmbedding`
* `IsInducing.isLindelof_iff` → `Topology.IsInducing.isLindelof_iff`
* `IsInducing.isLindelof_preimage` → `Topology.IsInducing.isLindelof_preimage`
* `IsInducing.isLocallyClosed_iff` → `Topology.IsInducing.isLocallyClosed_iff`
* `IsInducing.isOpenMap` → `Topology.IsInducing.isOpenMap`
* `IsInducing.isPathConnected_iff` → `Topology.IsInducing.isPathConnected_iff`
* `IsInducing.isPreconnected_image` → `Topology.IsInducing.isPreconnected_image`
* `IsInducing.isSigmaCompact_iff` → `Topology.IsInducing.isSigmaCompact_iff`
* `IsInducing.joinedIn_image` → `Topology.IsInducing.joinedIn_image`
* `IsInducing.locallyCompactSpace` → `Topology.IsInducing.locallyCompactSpace`
* `IsInducing.multipliable_iff_tprod_comp_mem_range` → `Topology.IsInducing.multipliable_iff_tprod_comp_mem_range`
* `IsInducing.of_codRestrict` → `Topology.IsInducing.of_codRestrict`
* `IsInducing.prodMap` → `Topology.IsInducing.prodMap`
* `IsInducing.r0Space` → `Topology.IsInducing.r0Space`
* `IsInducing.r1Space` → `Topology.IsInducing.r1Space`
* `IsInducing.regularSpace` → `Topology.IsInducing.regularSpace`
* `IsInducing.restrictPreimage` → `Topology.IsInducing.restrictPreimage`
* `IsInducing.secondCountableTopology` → `Topology.IsInducing.secondCountableTopology`
* `IsInducing.specializes_iff` → `Topology.IsInducing.specializes_iff`
* `IsInducing.specializingMap` → `Topology.IsInducing.specializingMap`
* `IsInducing.subtypeVal` → `Topology.IsInducing.subtypeVal`
* `IsInducing.topologicalGroup` → `Topology.IsInducing.topologicalGroup`
* `IsInducing.withSeminorms` → `Topology.IsInducing.withSeminorms`
* `IsOpenEmbedding.coborder_preimage` → `Topology.IsOpenEmbedding.coborder_preimage`
* `IsOpenEmbedding.compatiblePreserving` → `Topology.IsOpenEmbedding.compatiblePreserving`
* `IsOpenEmbedding.functor_isContinuous` → `Topology.IsOpenEmbedding.functor_isContinuous`
* `IsOpenEmbedding.functor_obj_injective` → `Topology.IsOpenEmbedding.functor_obj_injective`
* `IsOpenEmbedding.inl` → `Topology.IsOpenEmbedding.inl`
* `IsOpenEmbedding.inr` → `Topology.IsOpenEmbedding.inr`
* `IsOpenEmbedding.isLocalHomeomorph` → `Topology.IsOpenEmbedding.isLocalHomeomorph`
* `IsOpenEmbedding.isQuasiSeparated_iff` → `Topology.IsOpenEmbedding.isQuasiSeparated_iff`
* `IsOpenEmbedding.locPathConnectedSpace` → `Topology.IsOpenEmbedding.locPathConnectedSpace`
* `IsOpenEmbedding.locallyCompactSpace` → `Topology.IsOpenEmbedding.locallyCompactSpace`
* `IsOpenEmbedding.locallyConnectedSpace` → `Topology.IsOpenEmbedding.locallyConnectedSpace`
* `IsOpenEmbedding.map_nhdsWithin_preimage_eq` → `Topology.IsOpenEmbedding.map_nhdsWithin_preimage_eq`
* `IsOpenEmbedding.measurableEmbedding` → `Topology.IsOpenEmbedding.measurableEmbedding`
* `IsOpenEmbedding.preimage_closedPoints` → `Topology.IsOpenEmbedding.preimage_closedPoints`
* `IsOpenEmbedding.prodMap` → `Topology.IsOpenEmbedding.prodMap`
* `IsOpenEmbedding.quasiSober` → `Topology.IsOpenEmbedding.quasiSober`
* `IsOpenEmbedding.restrictPreimage` → `Topology.IsOpenEmbedding.restrictPreimage`
* `IsOpenEmbedding.sigmaMk` → `Topology.IsOpenEmbedding.sigmaMk`
* `IsOpenEmbedding.singleton_smoothManifoldWithCorners` → `Topology.IsOpenEmbedding.singleton_smoothManifoldWithCorners`
* `IsQuotientMap.continuousOn_isOpen_iff` → `Topology.IsQuotientMap.continuousOn_isOpen_iff`
* `IsQuotientMap.continuous_lift_prod_left` → `Topology.IsQuotientMap.continuous_lift_prod_left`
* `IsQuotientMap.continuous_lift_prod_right` → `Topology.IsQuotientMap.continuous_lift_prod_right`
* `IsQuotientMap.image_connectedComponent` → `Topology.IsQuotientMap.image_connectedComponent`
* `IsQuotientMap.isClopen_preimage` → `Topology.IsQuotientMap.isClopen_preimage`
* `IsQuotientMap.preimage_connectedComponent` → `Topology.IsQuotientMap.preimage_connectedComponent`
* `IsQuotientMap.restrictPreimage_isOpen` → `Topology.IsQuotientMap.restrictPreimage_isOpen`
* `IsQuotientMap.sequentialSpace` → `Topology.IsQuotientMap.sequentialSpace`
* `isEmbedding_sigmaMap` → `Topology.isEmbedding_sigmaMap`
* `isInducing_const_prod` → `+ Topology.isInducing_const_prod`
* `isInducing_prod_const` → `+ Topology.isInducing_prod_const`
* `isInducing_sigmaMap` → `+ Topology.isInducing_sigmaMap`
* `isOpenEmbedding_sigmaMap` → `+ Topology.isOpenEmbedding_sigmaMap`
* `IsClosedEmbedding.integral_map` → `+ Topology.IsClosedEmbedding.integral_map`
* `IsClosedEmbedding.polishSpace` → `Topology.IsClosedEmbedding.polishSpace`
* `IsClosedEmbedding.setIntegral_map` → `Topology.IsClosedEmbedding.setIntegral_map`
* `IsEmbedding.aestronglyMeasurable_comp_iff` → `Topology.IsEmbedding.aestronglyMeasurable_comp_iff`
* `IsEmbedding.firstCountableTopology` → `Topology.IsEmbedding.firstCountableTopology`
* `IsEmbedding.metrizableSpace` → `Topology.IsEmbedding.metrizableSpace`
* `IsEmbedding.toHomeomorph_of_surjective` → `Topology.IsEmbedding.toHomeomorph_of_surjective`
* `IsInducing.firstCountableTopology` → `Topology.IsInducing.firstCountableTopology`
* `IsInducing.noetherianSpace` → `Topology.IsInducing.noetherianSpace`
* `IsInducing.pseudoMetrizableSpace` → `Topology.IsInducing.pseudoMetrizableSpace`
* `IsQuotientMap.secondCountableTopology` → `Topology.IsQuotientMap.secondCountableTopology`
* `IsQuotientMap.separableSpace` → `Topology.IsQuotientMap`

Ruben-VandeVelde committed:
chore: move results about the kernel of Polynomial.modByMonicHom earlier (#18883)

These need no further imports, and are the only results that use Polynomial.modByMonicHom (or anything else from Algebra.Polynomial.RingDivision) in RingTheory.Polynomial.Basic.

j-loreaux committed:
feat: topological closure of non-unital subobjects (#18613)

This adds `*.topologicalClosure` and the related fundamental lemmas for each of `Semigroup`, `AddSemigroup`, `NonUnitalSemiring`, `NonUnitalSubring`, `NonUnitalAlgebra` and `NonUnitalStarAlgebra`, thereby filling a gap in our API for these objects.

adomani committed:
CI(maintainer merge): emphasize merge/delegate (#18890)

Makes the words `delegate` and `merge` bold in the Zulip post triggered by `maintainer delegate`/`maintainer merge`.

winstonyin committed:
chore: split `IntegralCurve` (#18833)

Split `Mathlib.Geometry.Manifold.IntegralCurve` into three files:
* `Mathlib.Geometry.Manifold.IntegralCurve.Basic`: Definitions and basic lemmas about continuity and derivatives of integral curves
* `Mathlib.Geometry.Manifold.IntegralCurve.Transform`: Lemmas about translation and scaling of the domain of integral curves by a constant
* `Mathlib.Geometry.Manifold.IntegralCurve.ExistUnique`: Local existence and uniqueness theorems for integral curves

No content has been substantively changed from the original file. This is to prepare for #9013, which proves a sufficient condition for global existence of integral curves. That'll go in another new file.

vihdzp committed:
refactor(SetTheory/Cardinal/Cofinality): deprecate `StrictOrder.cof` (#18043)

This definition is incorrect for partial orders and redundant for linear orders, see [Zulip](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/RelIso.2Ecof_eq).

mathlib-bors[bot] committed:
chore: update Mathlib dependencies 2024-11-12 (#18897)

This PR updates the Mathlib dependencies.

kim-em committed:
chore: cleanup of simpNF (#18864)

Lots more to go.

vihdzp committed:
chore(Order/InitialSeg): fix simp lemmas (#17602)

See [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/simp.20and.20subrel).

mathlib-bors[bot] committed:
chore: update Mathlib dependencies 2024-11-12 (#18900)

This PR updates the Mathlib dependencies.

vihdzp committed:
chore: `Data/List/Join` → `Data/List/Flatten` (#18895)

TpmKranz committed:
feat(Data/FinEnum): Option instance and Type recursor (#15647)

Replicate `Data/Fintype/Option`, but yield non-truncated data.

First chunk of #12648

kmill committed:
fix: multigoal linter should not flag tactics that do nothing (#18817)

When a tactic does not change the list of goals, the multigoal linter was giving incorrect error messages such as
```
There are 2 unclosed goals before 'change ...' and
at least one remaining goal afterwards.
Please focus on the current goal, for instance
using `·` (typed as "\.").
```
This PR adjusts the condition to allow all tactics that operate on one goal and result in zero or more goals. It also makes the error message more precise, for example
```
warning: The following tactic starts with 2 goals and
ends with 1 goal, 1 of which is not operated on.
  exact .intro
Please focus on the current goal, for instance
using `·` (typed as "\.").
```




Co-authored-by: Damiano Testa <adomani@gmail.com>

ooovi committed:
feat(Combinatorics/SimpleGraph/Clique): add clique number (#18446)

add clique number, maximal and maximum cliques, and theorems about the relationship between clique number and the cardinality of cliques.



Co-authored-by: Christoph Spiegel <christophspiegel.berlin@gmail.comm>
Co-authored-by: FordUniver <61389961+FordUniver@users.noreply.github.com>
Co-authored-by: ooovi <79147175+ooovi@users.noreply.github.com>

xroblot committed:
feat(MeasureTheory): prove that some equivalences are measure preserving (#17944)

Construct measurable equivalence for the following equiv and then prove that they are measure preserving: 
- `Equiv.arrowProdEquivProdArrow`
- `Equiv.arrowCongr'`
- `Equiv.prodAssoc`

Also, we fix a typo in [MeasureTheory.measurePreserving_pi](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Constructions/Pi.html#MeasureTheory.measurePreserving_pi)

This PR is part of the proof of the Analytic Class Number Formula.

vihdzp committed:
chore(SetTheory/Ordinal/Principal): use `variable` (#18898)

...plus a simple golf for `isLimit_of_principal_add`.

YaelDillies committed:
chore: rename `LocalRing` to `IsLocalRing` (#18774)

Co-authored-by: Javier López-Contreras <jlopezcontreras10@gmail.com>

Ruben-VandeVelde committed:
chore: fix the name of the IsNClique.isClique field (#18908)

Followup to #18446

mathlib-bors[bot] committed:
chore: update Mathlib dependencies 2024-11-12 (#18906)

This PR updates the Mathlib dependencies.

rwst committed:
feat(Nat/Prime): Primality of some numbers of form `a^n±1` (#14874)

New file `Nat/Prime/MersenneFermat.lean` contains the theorems: 

- Prime `a ^ n + 1` implies `n` is a power of two.
- `Fₙ = 2^(2^n)+1` is prime only if `3^(2^(2^n-1)) = -1 mod Fₙ` (Pépin's test).
- Prime factors of `a ^ (2 ^ n) + 1` are of form `k * 2 ^ (n + 1) + 1`.
- Prime factors of `Fₙ = 2 ^ (2 ^ n) + 1`, `1 < n`, are of form `k * 2 ^ (n + 2) + 1`.
- Prime factors of `Fₙ = 2 ^ (2 ^ n) + 1` are either 3, 5, or of form `k * 2 ^ (n + 2) + 1`.
- Prime `a ^ n - 1` implies `a = 2` and prime `n`.

Needs several imports from other modules so it was not wise to add it to `Prime/Basic`. 

    Co-authored-by: Thomas Browning <tb65536@uw.edu>



Co-authored-by: Moritz Firsching <firsching@google.com>

ADedecker committed:
chore(Analysis.Polynomial.Basic): fix name in module docstring (#18909)

j-loreaux committed:
feat: make `tsub_self` a `simp` lemma (#18871)

The lemma `tsub_eq_zero_iff_le` used to be `@[simp]`, but this was (intentionally) changed a few months ago. As a result, goals like `(1 : ℝ≥0) - 1 = 0` are no longer provable by any easy tactic, which is silly. The solution is to mark `tsub_self` with `@[simp]`, which is safe.

chrisflav committed:
feat(AlgebraicGeometry/Morphisms): finite morphisms (#18357)

This PR defines finite morphisms of schemes.

b-mehta committed:
feat(Algebra/PUnitInstances): every set has a PUnit-action (#18614)

adomani committed:
chore: more unused variables (#18889)

#17715

vihdzp commented on Mathlib/Data/Sym/Sym2.lean:
```diff
@@ -566,6 +572,48 @@ def sym2EquivSym' : Equiv (Sym2 α) (Sym' α 2) where
+  mem_equivSym_symm_iff_mem_of_eq rfl
+
+theorem equivSym_map_comm {f : α → β} {s : Sym2 α} :
+    (Sym2.equivSym β) (s.map f) = (Sym2.equivSym _ s).map f := by
```
```suggestion
    Sym2.equivSym β (s.map f) = (Sym2.equivSym _ s).map f := by
```

vihdzp commented on Mathlib/Data/Sym/Sym2.lean:
```diff
@@ -566,6 +572,48 @@ def sym2EquivSym' : Equiv (Sym2 α) (Sym' α 2) where
+  rfl
+
+theorem IsDiag.equivSym_symm_replicate (a : α) :
+    ((equivSym α).symm (Sym.replicate 2 a)).IsDiag  := by
```
```suggestion
    ((equivSym α).symm (Sym.replicate 2 a)).IsDiag := by
```

vihdzp commented on Mathlib/Data/Sym/Sym2.lean:
```diff
@@ -566,6 +572,48 @@ def sym2EquivSym' : Equiv (Sym2 α) (Sym' α 2) where
+  simp only [replicate, Multiset.replicate_succ, Multiset.replicate_zero, Multiset.cons_zero,
+    equivSym_symm_eq_pair, isDiag_iff_proj_eq]
+
+theorem equivSym_IsDiag_eq_replicate {s : Sym2 α} :
```
```suggestion
theorem isDiag_iff_equivSym_eq_replicate {s : Sym2 α} :
```

vihdzp reviewed (commented):
some style stuff

Jun2M committed:
Update Mathlib/Data/Sym/Sym2.lean

Co-authored-by: Violeta Hernández <vi.hdz.p@gmail.com>

Jun2M committed:
Update Mathlib/Data/Sym/Sym2.lean

Co-authored-by: Violeta Hernández <vi.hdz.p@gmail.com>

Jun2M committed:
Update Mathlib/Data/Sym/Sym2.lean

Co-authored-by: Violeta Hernández <vi.hdz.p@gmail.com>

Jun2M commented:
Oh, no. what have I done.
